libc-bottom-half/cloudlibc/src/libc/sys/wasi_preview2/descriptor_table.c

```diff
@@ -0,0 +1,246 @@
+/*
+ * This file provides a global hashtable for tracking `wasi-libc`-managed file
+ * descriptors.
+ *
+ * WASI Preview 2 has no notion of file descriptors and instead uses unforgeable
+ * resource handles.  Moreover, there's not necessarily a one-to-one
+ * correspondence between POSIX file descriptors and resource handles (e.g. a
+ * TCP connection may require separate handles for reading, writing, and polling
+ * the same connection).  We use this table to map each POSIX descriptor to a
+ * set of one or more handles.
+ *
+ * As of this writing, we still rely on the WASI Preview 1 adapter
+ * (https://github.com/bytecodealliance/wasmtime/tree/main/crates/wasi-preview1-component-adapter)
+ * to manage non-socket descriptors, so currently this table only tracks TCP and
+ * UDP sockets.  We use the adapter's `adapter_open_badfd` and
+ * `adapter_close_badfd` functions to reserve and later close descriptors to
+ * avoid confusion (e.g. if an application tries to use Preview 1 host functions
+ * directly for socket operations rather than go through `wasi-libc`).
+ * Eventually, we'll switch `wasi-libc` over to Preview 2 entirely, at which
+ * point we'll no longer need the adapter.  At that point, all file descriptors
+ * will be managed exclusively in this table.
+ */
+
+#include <assert.h>
+
+#include <descriptor_table.h>
+
+__attribute__((__import_module__("wasi_snapshot_preview1"), __import_name__("adapter_open_badfd"))) extern int32_t __wasi_preview1_adapter_open_badfd(int32_t);
```

Comment:
These imports are very much not meant to be implemented in hosts. Or rather, they're absolutely not meant to become part of preview1, so naming them in a way that makes them seem as though they are would be wrong.

Reply:
> These imports are very much not meant to be implemented in hosts.

why not?
after all, the adapter can be considered as a part of the "host" from the POV of this module, can't it?

Reply:
Yes, but that doesn't make them part of the wasi-preview1 standard. The goal here is to make all of this internal details of wasi-sdk, where any version might move functionality from the adapter into direct use of preview2 APIs. If hosts were to implement these APIs, that means they break compatibility by relying on SDK-internal implementation details.

Reply:
any version can work if a host provides a complete preview1 implementation + this badfd api + enough coverage of preview2 implementation, can't it?

Reply:
the goal of preview2 support is to create a wasi-libc target for preview2 hosts, not for hosts that implement a combination of preview1 APIs plus some preview2 APIs plus some non-standard APIs. As such, we absolutely should not encourage hosts to implement this kind of mixture of APIs to make some sort of random version of wasi-libc work because the implementation details happen to align.